### DISCUSSION TITLE
porting project to new vivado version(2021.2)

### DISCUSSION THREAD

**Discussion by julianctrt** [srsran/zynq_timestamping#33] [General]
this repo is only designed for the 2019.2 version and due my toolchain is the 2021.2 version I would like update all code.

for now the vivado project is updated and generated the bitstream for the zcu208 dev board but not sure if it's working since I can't still test it because I'm still not able to test the txrx app for not been able to configure the kernel.

After generate the petalinux image and boot it in the dev board correctly the next step is generate the environment for cross compile the app.

I understand I have to checkout the 2021.2 version of the linux-xlnx repo since the petalinux was created with this version.

but the config is not working correctly since is asking for lot of configuration so I guess some info is missing in config file.

could you provide the complete config for compare it?.

regards

**Comment by ofontbach** [comment]
Hi @julianctrt,

it's great to see that you've advanced in your idea described #30. Once it works, it would be nice to share the details of how you have modified the solution and, if you like, the code too. Please, let us know if we can help you towards that end.

As you said, we used version 2019.2 for both Vivado and Petalinux for the Zcu boards. We have in our short-term plans to move to newer versions. In the meanwhile, what is exactly the configuration information that you need? Let us know and we'll share it here. Thanks!

**Reply by julianctrt** [reply]
hello @sauka and @ofontbach  and thanks for reply.

in the documentation you provide for the kernel:

[](https://srsran.github.io/zynq_timestamping/app/zcu.html#compiling-a-customized-linux-kernel-from-source-code)

I have problem with this part:

```
scp -r root@<zcu111-ip-address>:/proc/config.gz .
zcat config.gz > .config
```

I was able to generate the petalinux with all the configuration mention before and run in my zcu208 board but, for now I was only able to make it works for the prebuild version of petalinux.

after downloading the config.gz and extracting to my linux-xlnx folder when run the  command:

`make ARCH=arm64 oldconfig`

a lot question about kernel configuration prompt in the console like the configuration is missing them.

maybe if you provide this file I can compare it with mine and see what's happening.

also if I ignore them and set as default I got errors when running the make for the compiled_modules (first make finish without errors):

```
make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- -j8 INSTALL_MOD_PATH=./compiled_modules modules
scripts/kconfig/conf  --syncconfig Kconfig
  CC      kernel/bounds.s
  CALL    scripts/atomic/check-atomics.sh
  CC      arch/arm64/kernel/asm-offsets.s
  CALL    scripts/checksyscalls.sh
  AS      arch/arm64/kernel/vdso/sigreturn.o
  CC      arch/arm64/kernel/vdso/vgettimeofday.o
  LD      arch/arm64/kernel/vdso/vdso.so.dbg
  VDSOSYM include/generated/vdso-offsets.h
  CC [M]  drivers/phy/ti/phy-gmii-sel.o
  Building modules, stage 2.
  MODPOST 1 modules
ERROR: "platform_driver_unregister" [drivers/phy/ti/phy-gmii-sel.ko] undefined!
ERROR: "__platform_driver_register" [drivers/phy/ti/phy-gmii-sel.ko] undefined!
ERROR: "_dev_err" [drivers/phy/ti/phy-gmii-sel.ko] undefined!
ERROR: "devm_kmalloc" [drivers/phy/ti/phy-gmii-sel.ko] undefined!
ERROR: "of_match_node" [drivers/phy/ti/phy-gmii-sel.ko] undefined!
make[1]: *** [scripts/Makefile.modpost:94: __modpost] Error 1
make: *** [Makefile:1285: modules] Error 2
```
but this could be maybe to various reason so at first I want to discard the config file.

Regards

**Reply by sauka** [reply]
Thanks for the details. Could you try checking out tag xilinx-v2021.2 and just compile it with a default config? I.e. don't use config.gz and instead of running
`make ARCH=arm64 oldconfig`
try
`make ARCH=arm64 xilinx_zynqmp_defconfig`

**Reply by sauka** [reply]
Could you also share the output of `uname -a` from a running Linux kernel built by Petalinux tools?
I just checked that `xilinx_zynqmp_defconfig` compiles without problem. I wonder if the kernel version used by petalinux is different.

**Reply by julianctrt** [reply]
Hello @sauka and thanks for the support,

Tried the command and it compiles, at least I was able to run the commands without errors:

```
make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- -j8 Image UIMAGE_LOADADDR=0x8000
mkdir compiled_modules
make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- -j8 INSTALL_MOD_PATH=./compiled_modules modules
make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- -j8 INSTALL_MOD_PATH=./compiled_modules modules_install
  INSTALL crypto/crypto_engine.ko
  INSTALL drivers/crypto/virtio/virtio_crypto.ko
  INSTALL drivers/media/usb/gspca/gspca_main.ko
  INSTALL drivers/mtd/tests/mtd_nandbiterrs.ko
  INSTALL drivers/mtd/tests/mtd_nandecctest.ko
  INSTALL drivers/mtd/tests/mtd_oobtest.ko
  INSTALL drivers/mtd/tests/mtd_pagetest.ko
  INSTALL drivers/mtd/tests/mtd_readtest.ko
  INSTALL drivers/mtd/tests/mtd_speedtest.ko
  INSTALL drivers/mtd/tests/mtd_stresstest.ko
  INSTALL drivers/mtd/tests/mtd_subpagetest.ko
  INSTALL drivers/mtd/tests/mtd_torturetest.ko
  INSTALL drivers/remoteproc/zynqmp_r5_remoteproc.ko
  INSTALL drivers/rpmsg/rpmsg_core.ko
  INSTALL drivers/rpmsg/virtio_rpmsg_bus.ko
  INSTALL drivers/soc/xilinx/xlnx_vcu.ko
  INSTALL drivers/soc/xilinx/xlnx_vcu_clk.ko
  INSTALL drivers/soc/xilinx/xlnx_vcu_core.ko
  INSTALL drivers/uio/uio_dmem_genirq.ko
  INSTALL drivers/uio/uio_pdrv_genirq.ko
  INSTALL drivers/usb/gadget/function/usb_f_ecm.ko
  INSTALL drivers/usb/gadget/function/usb_f_ecm_subset.ko
  INSTALL drivers/usb/gadget/function/usb_f_rndis.ko
  INSTALL drivers/usb/gadget/legacy/g_ether.ko
  INSTALL drivers/usb/gadget/legacy/g_mass_storage.ko
  INSTALL fs/efivarfs/efivarfs.ko
  INSTALL fs/nfs/flexfilelayout/nfs_layout_flexfiles.ko
  INSTALL net/bridge/br_netfilter.ko
  INSTALL net/ipv4/netfilter/nf_defrag_ipv4.ko
  INSTALL net/netfilter/nf_conntrack.ko
  INSTALL net/ipv6/netfilter/nf_defrag_ipv6.ko
  INSTALL net/netfilter/nf_conntrack_netlink.ko
  INSTALL net/netfilter/xt_connmark.ko
  INSTALL net/netfilter/xt_conntrack.ko
  INSTALL net/netfilter/xt_state.ko
```

By the way, mi uname -a is: 

`Linux julian-GF65-Thin-9SD 5.15.0-56-generic #62~20.04.1-Ubuntu SMP Tue Nov 22 21:24:20 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`

First doubt I have, in the documentation it says:

_"Copy the modules to /lib/modules/ in the rootfs partition of the card"_

In my card I have a folder /lib/modules/5.10.0-xilinx-v2021.2 but the generated compiled_modules is compiled_modules/lib/modules/5.4.0

Should I overwrite the content of 5.4.0 in 5.10.0-xilinx-v2021.2?

Also if continue with the  process to be able to run the txrx example after Compiling a custom Linux kernel driver chapter I have problems in the last step Cross-compiling the Zynq timestamping library and Tx-Rx example application.

After install the sdk I have problem with the prepare.sh:

```
./prepare.sh rfsoc
mkdir: cannot create directory ‘/home/julian/Documents/RFsoc/zynq_timestamping/app/bin_app’: File exists
--> Current UE configuration
• Board IP: 10.12.1.201
• freq_offset: -4100
• rx_gain: 50
• time_adv_nsamples: 82

mkdir: cannot create directory ‘build_arm’: File exists
Preparing for cross-compilation
-- CMAKE_SYSTEM: Linux
-- CMAKE_SYSTEM_PROCESSOR: cortexa72-cortexa53
-- CMAKE_CXX_COMPILER: /home/julian/Documents/RFsoc/xilinx-zcu208-2021.2/images/linux/sysroots/x86_64-petalinux-linux/usr/bin/aarch64-xilinx-linux/aarch64-xilinx-linux-g++
-- Checking for module 'fftw3f >= 3.0'
--   No package 'fftw3f' found
-- FFTW3F LIBRARIES: FFTW3F_LIBRARY-NOTFOUND
-- FFTW3F STATIC LIBRARIES: FFTW3F_STATIC_LIBRARY-NOTFOUND
-- FFTW3F INCLUDE DIRS: FFTW3F_INCLUDE_DIR-NOTFOUND
CMake Error at /home/julian/Documents/RFsoc/xilinx-zcu208-2021.2/images/linux/sysroots/x86_64-petalinux-linux/usr/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
  Could NOT find FFTW3F (missing: FFTW3F_LIBRARY FFTW3F_INCLUDE_DIR)
Call Stack (most recent call first):
  /home/julian/Documents/RFsoc/xilinx-zcu208-2021.2/images/linux/sysroots/x86_64-petalinux-linux/usr/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindFFTW3F.cmake:61 (find_package_handle_standard_args)
  CMakeLists.txt:169 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/julian/Documents/RFsoc/zynq_timestamping/app/srsRAN/build_arm/CMakeFiles/CMakeOutput.log".
See also "/home/julian/Documents/RFsoc/zynq_timestamping/app/srsRAN/build_arm/CMakeFiles/CMakeError.log".
make: *** No rule to make target 'usrp_txrx'.  Stop.
Copying configuration files...
cp: cannot stat 'lib/examples/usrp_txrx': No such file or directory
cp: cannot stat 'lib/examples/pdsch_ue': No such file or directory
-- CMAKE_SYSTEM: Linux
-- CMAKE_SYSTEM_PROCESSOR: cortexa72-cortexa53
-- CMAKE_CXX_COMPILER: /home/julian/Documents/RFsoc/xilinx-zcu208-2021.2/images/linux/sysroots/x86_64-petalinux-linux/usr/bin/aarch64-xilinx-linux/aarch64-xilinx-linux-g++
-- Could NOT find libiio (missing: LIBIIO_LIBRARY LIBIIO_INCLUDE_DIR) (Required is at least version "0.9")
CMake Error at CMakeLists.txt:73 (message):
  libiio not found


-- Configuring incomplete, errors occurred!
See also "/home/julian/Documents/RFsoc/zynq_timestamping/sw/build_arm/CMakeFiles/CMakeOutput.log".
See also "/home/julian/Documents/RFsoc/zynq_timestamping/sw/build_arm/CMakeFiles/CMakeError.log".
make: *** No targets specified and no makefile found.  Stop.
cp: cannot stat './install/lib/*': No such file or directory
Done
```

First, the "rfsoc" parameter is not a valid parameter as far I can understand in the prepare.sh and should be "zcu111" instead of "rfsoc":

```
board_name=$1
if [[ "$board_name" == "zcu111" ]]; then
  cmake_rf_cfg="-DENABLE_IIO=Off -DENABLE_RFDC=On"
else
  cmake_rf_cfg="-DENABLE_IIO=On -DENABLE_RFDC=Off"
fi
```

But if I run _prepare.sh zcu111_ also get similar error:

```
./prepare.sh zcu111
mkdir: cannot create directory ‘/home/julian/Documents/RFsoc/zynq_timestamping/app/bin_app’: File exists
--> Current UE configuration
• Board IP: 10.12.1.201
• freq_offset: -4100
• rx_gain: 50
• time_adv_nsamples: 82

mkdir: cannot create directory ‘build_arm’: File exists
Preparing for cross-compilation
-- CMAKE_SYSTEM: Linux
-- CMAKE_SYSTEM_PROCESSOR: cortexa72-cortexa53
-- CMAKE_CXX_COMPILER: /home/julian/Documents/RFsoc/xilinx-zcu208-2021.2/images/linux/sysroots/x86_64-petalinux-linux/usr/bin/aarch64-xilinx-linux/aarch64-xilinx-linux-g++
-- Checking for module 'fftw3f >= 3.0'
--   No package 'fftw3f' found
-- FFTW3F LIBRARIES: FFTW3F_LIBRARY-NOTFOUND
-- FFTW3F STATIC LIBRARIES: FFTW3F_STATIC_LIBRARY-NOTFOUND
-- FFTW3F INCLUDE DIRS: FFTW3F_INCLUDE_DIR-NOTFOUND
CMake Error at /home/julian/Documents/RFsoc/xilinx-zcu208-2021.2/images/linux/sysroots/x86_64-petalinux-linux/usr/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
  Could NOT find FFTW3F (missing: FFTW3F_LIBRARY FFTW3F_INCLUDE_DIR)
Call Stack (most recent call first):
  /home/julian/Documents/RFsoc/xilinx-zcu208-2021.2/images/linux/sysroots/x86_64-petalinux-linux/usr/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindFFTW3F.cmake:61 (find_package_handle_standard_args)
  CMakeLists.txt:169 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/julian/Documents/RFsoc/zynq_timestamping/app/srsRAN/build_arm/CMakeFiles/CMakeOutput.log".
See also "/home/julian/Documents/RFsoc/zynq_timestamping/app/srsRAN/build_arm/CMakeFiles/CMakeError.log".
make: *** No rule to make target 'usrp_txrx'.  Stop.
Copying configuration files...
cp: cannot stat 'lib/examples/usrp_txrx': No such file or directory
cp: cannot stat 'lib/examples/pdsch_ue': No such file or directory
-- CMAKE_SYSTEM: Linux
-- CMAKE_SYSTEM_PROCESSOR: cortexa72-cortexa53
-- CMAKE_CXX_COMPILER: /home/julian/Documents/RFsoc/xilinx-zcu208-2021.2/images/linux/sysroots/x86_64-petalinux-linux/usr/bin/aarch64-xilinx-linux/aarch64-xilinx-linux-g++
-- libmetal library found
--   LIBMETAL_INCLUDE_DIR: /home/julian/Documents/RFsoc/xilinx-zcu208-2021.2/images/linux/sysroots/cortexa72-cortexa53-xilinx-linux/usr/include
--   LIBMETAL_LIBRARIES: /home/julian/Documents/RFsoc/xilinx-zcu208-2021.2/images/linux/sysroots/cortexa72-cortexa53-xilinx-linux/usr/lib/libmetal.so
-- RFDC library found
--   RFDC_INCLUDE_DIR: /home/julian/Documents/RFsoc/xilinx-zcu208-2021.2/images/linux/sysroots/cortexa72-cortexa53-xilinx-linux/usr/include
--   RFDC_LIBRARY: /home/julian/Documents/RFsoc/xilinx-zcu208-2021.2/images/linux/sysroots/cortexa72-cortexa53-xilinx-linux/usr/lib/librfdc.so
-- /home/julian/Documents/RFsoc/zynq_timestamping/sw/build_arm/install/lib does not exist. Defaulting install location to /home/julian/Documents/RFsoc/zynq_timestamping/sw/build_arm/install/lib.
-- Performing Test HAVE_WNO_UNUSED_BUT_SET_VARIABLE
-- Performing Test HAVE_WNO_UNUSED_BUT_SET_VARIABLE - Success
-- Performing Test HAVE_SSE
-- Performing Test HAVE_SSE - Failed
-- Performing Test HAVE_AVX
-- Performing Test HAVE_AVX - Failed
-- Performing Test HAVE_AVX2
-- Performing Test HAVE_AVX2 - Failed
-- Performing Test HAVE_FMA
-- Performing Test HAVE_FMA - Failed
-- Performing Test HAVE_AVX512
-- Performing Test HAVE_AVX512 - Failed
-- Performing Test HAVE_MARCH_native
-- Performing Test HAVE_MARCH_native - Failed
-- Performing Test HAVE_VISIBILITY_HIDDEN_CXX
-- Performing Test HAVE_VISIBILITY_HIDDEN_CXX - Failed
-- Performing Test HAVE_ERROR_INCOMPATIBLE
-- Performing Test HAVE_ERROR_INCOMPATIBLE - Failed
CMake Error at CMakeLists.txt:247 (message):
  no SIMD instructions found


-- Configuring incomplete, errors occurred!
See also "/home/julian/Documents/RFsoc/zynq_timestamping/sw/build_arm/CMakeFiles/CMakeOutput.log".
See also "/home/julian/Documents/RFsoc/zynq_timestamping/sw/build_arm/CMakeFiles/CMakeError.log".
make: *** No targets specified and no makefile found.  Stop.
cp: cannot stat './install/lib/*': No such file or directory
Done

```

But this errors are related with the FFTW3F library since it seems or there is a problem with the version check(I have the 3.5.8 and is asking for a very old one 3.0) or the library paths are not located.

What should I do? going for the old version or remove the check version to compile it with the new fftw library?(don't know where is the make file who checks the fftw version)

Regards

**Reply by sauka** [reply]
Hi @julianctrt 

Good catch about `prepare.sh` script. It is a mistake in the appnote,  as you found out it should be `./prepare.sh zcu111` (in the [README](https://github.com/srsran/zynq_timestamping/tree/main/app) the command looks correct),  Could you open an issue for this?

As for building the linux-xlnx kernel manually from sources with default xilinx_zynqmp_defconfig config, I see it compiles well and installs modules to 5.10.0, i.e.
```
pavel@pavel-XPS-13-9310:/tmp/linux-xlnx$ tree compiled_modules/
compiled_modules/
└── lib
    └── modules
        └── 5.10.0
```
Might be that wrong branch was used and you compiled kernel 5.4.0?

**Reply by sauka** [reply]
regarding FFTW, CMake searches for it (version 3.0 or newer) in your sysroot under `sysroots/cortexa72-cortexa53-xilinx-linux`. Are you sure you have it there? If not you can ask Petalinux tools to add fftw to your image and respectively to the sysroot (the appnote describes this step).

p.s. it is true that for txrx test fftw shouldn't be needed, but this is marked as required by srsRAN submodule, so we have to fulfill this requirement

**Reply by julianctrt** [reply]
thanks @sauka for the info,

the install modules was solved and now I have the 5.10 version.

in the other side the FFTW is still missing since I'm not able to find the petalinux-user-image.bbappend to add the extra dependencies since the recipes-core is not present due in newer version of petalinux from 2019.2 the method to include additional dependencies is done in other way:

https://support.xilinx.com/s/question/0D52E00006hpKuRSAU/petalinux-need-metauserrecipescore-directory-in-project-to-add-a-userpackage-to-rootfs-but-the-directory-doesnt-exist?language=en_US

and I didn't do it since from documentation it says:

"For the ZCU111 build case, no extra dependencies should be needed to cross-compile the Zynq timestamping software and txrx example application"

but since It will needed for SRS I guess I'll go for it from the beginning I'll try to rebuild including the extra dependencies. 

regards

P.D in my petalinux config this menu are missing:

. Subsystem AUTO Hardware Settings ->
    .. Advanced bootable images storage Settings ->
        * boot image settings    -> image storage media -> primary sd
        * kernel image settings  -> image storage media -> primary sd
                                 -> image name -> Image
        * dtb image settings     -> image storage media -> primary sd
. Yocto Settings ->
    .. Enable Debug Tweaks

and don't know why, any ideas?

**Comment by sauka** [comment]
Hi @julianctrt 

> I understand I have to checkout the 2021.2 version of the linux-xlnx repo since the petalinux was created with this version.
> 
> but the config is not working correctly since is asking for lot of configuration so I guess some info is missing in config file.

Could you clarify which config file are you using?
